testing/test_actors/actors/fil-upgrade-actor/src/actor.rs

```diff
@@ -0,0 +1,128 @@
+            let new_code_cid = sdk::actor::get_actor_code_cid(&Address::new_id(10000)).unwrap();
+            let params = IpldBlock::serialize_cbor(&SomeStruct { value: 4 }).unwrap();
+            let _ = sdk::actor::upgrade_actor(new_code_cid, params);
+            unreachable!("we should never return from a successful upgrade");
+        }
+        4 => {
+            let block_id = sdk::ipld::put_block(CBOR, &to_vec(&444).unwrap()).unwrap();
+            sdk::debug::log(format!(
+                "[upgrade] params:4, inside upgrade within an upgrade, returning block_id {}",
+                block_id
+            ));
+            block_id
+        }
+        other => {
+            panic!("unexpected value: {}", other);
+        }
+    }
+}
+
+#[no_mangle]
+pub fn invoke(_: u32) -> u32 {
+    sdk::initialize();
+
+    match sdk::message::method_number() {
+        // test that successful calls to `upgrade_actor` does not return
+        1 => {
+            let new_code_cid = sdk::actor::get_actor_code_cid(&Address::new_id(10000)).unwrap();
+            let params = IpldBlock::serialize_cbor(&SomeStruct { value: 1 }).unwrap();
+            let _ = sdk::actor::upgrade_actor(new_code_cid, params);
+            unreachable!("we should never return from a successful upgrade");
+        }
+        // test that when `upgrade` endpoint rejects upgrade that we get the returned exit code
+        2 => {
+            let new_code_cid = sdk::actor::get_actor_code_cid(&Address::new_id(10000)).unwrap();
+            let params = IpldBlock::serialize_cbor(&SomeStruct { value: 2 }).unwrap();
+            let res = sdk::actor::upgrade_actor(new_code_cid, params).unwrap();
+            assert_eq!(
+                UPGRADE_FAILED_EXIT_CODE,
+                res.exit_code.value(),
+                "invalid exit code returned from upgrade_actor"
+            );
+        }
+        // test recursive update
+        3 => {
+            let new_code_cid = sdk::actor::get_actor_code_cid(&Address::new_id(10000)).unwrap();
+            let params = IpldBlock::serialize_cbor(&SomeStruct { value: 3 }).unwrap();
+            let _ = sdk::actor::upgrade_actor(new_code_cid, params);
+            unreachable!("we should never return from a successful upgrade");
+        }
+        // test sending a message to ourself (putting us on the call stack)
+        4 => {
+            sdk::send::send(
+                &Address::new_id(10000),
+                5,
+                Default::default(),
+                TokenAmount::from_atto(100),
+                None,
+                Default::default(),
+            )
+            .unwrap();
+        }
+        // test that calling an upgrade with actor already on the call stack fails
+        5 => {
+            let new_code_cid = sdk::actor::get_actor_code_cid(&Address::new_id(10000)).unwrap();
+            let res = sdk::actor::upgrade_actor(new_code_cid, None);
+            assert_eq!(res, Err(ErrorNumber::Forbidden));
```

Comment:
Now that I think about this, I wonder if this should be `IllegalOperation`? That would let us distinguish between two possible failure cases.

Reply:
Not sure about changing it to `IllegalOperation` as that implies the actor is not in the correct state, which its not necessarily correct here

Reply:
Yeah, you're probably right. Anyways, now that I went back and looked at the code... we use a `SYS_INVALID_RECEIVER` exit code instead of a `Forbidden` error number to indicate that the target code doesn't implement the "upgrade" entrypoint. So there's no ambiguity anyways.